applications/thingy91/usb_uart_bridge/src/main.c

```diff
@@ -0,0 +1,284 @@
+ * Here, we target 64-byte blocks. Since we want to fit one struct uart_data
+ * into each block, the block layout becomes:
+ * 16 bytes: reserved by the Zephyr heap block descriptor (not in struct)
+ *  4 bytes: reserved by the Zephyr FIFO (in struct)
+ * 40 bytes: UART data buffer (in struct)
+ *  4 bytes: length field (in struct, padded for alignment)
+ */
+#define UART_BUF_SIZE 40
+
+static K_FIFO_DEFINE(usb_0_tx_fifo);
+static K_FIFO_DEFINE(usb_1_tx_fifo);
+static K_FIFO_DEFINE(uart_0_tx_fifo);
+static K_FIFO_DEFINE(uart_1_tx_fifo);
+
+struct uart_data {
+	void *fifo_reserved;
+	u8_t buffer[UART_BUF_SIZE];
+	u16_t len;
+};
+
+static struct serial_dev {
+	struct device *dev;
+	void *peer;
+	struct k_fifo *fifo;
+	struct k_sem sem;
+	struct uart_data *rx;
+} devs[4];
+
+/* Frees data for incoming transmission on device blocked by full heap. */
+int oom_free(struct serial_dev *sd)
+{
+	struct serial_dev *peer_sd = (struct serial_dev *)sd->peer;
+	struct uart_data *buf;
+
+	/* First, try to free from FIFO of peer device (blocked stream) */
+	buf = k_fifo_get(peer_sd->fifo, K_NO_WAIT);
+	if (buf) {
+		k_free(buf);
+		return 0;
+	}
+
+	/* Then, try FIFO of the receiving device (reverse of blocked stream) */
+	buf = k_fifo_get(sd->fifo, K_NO_WAIT);
+	if (buf) {
+		k_free(buf);
+		return 0;
+	}
+
+	/* Finally, try all of them */
+	for (int i = 0; i < ARRAY_SIZE(devs); i++) {
+		buf = k_fifo_get(sd->fifo, K_NO_WAIT);
+		if (buf) {
+			k_free(buf);
+			return 0;
+		}
+	}
+
+	return -1; /* Was not able to free any heap memory */
+}
+
+static void uart_interrupt_handler(void *user_data)
+{
+	struct serial_dev *sd = user_data;
+	struct device *dev = sd->dev;
+	struct serial_dev *peer_sd = (struct serial_dev *)sd->peer;
+
+	uart_irq_update(dev);
+
+	while (uart_irq_rx_ready(dev)) {
+		int data_length;
+
+		while (!sd->rx) {
+			sd->rx = k_malloc(sizeof(*sd->rx));
+			if (sd->rx) {
+				sd->rx->len = 0;
+			} else {
+				int err = oom_free(sd);
+
+				if (err) {
+					printk("Could not free memory. Rebooting.\n");
+					sys_reboot(SYS_REBOOT_COLD);
+				}
+			}
+		}
+
+		data_length = uart_fifo_read(dev, &sd->rx->buffer[sd->rx->len],
+					   UART_BUF_SIZE - sd->rx->len);
+		sd->rx->len += data_length;
+
+		if (sd->rx->len > 0) {
+			if ((sd->rx->len == UART_BUF_SIZE) ||
+			   (sd->rx->buffer[sd->rx->len - 1] == '\n') ||
+			   (sd->rx->buffer[sd->rx->len - 1] == '\r') ||
+			   (sd->rx->buffer[sd->rx->len - 1] == '\0')) {
+				k_fifo_put(peer_sd->fifo, sd->rx);
+				k_sem_give(&peer_sd->sem);
+
+				sd->rx = NULL;
+			}
+		}
+	}
+
+	if (uart_irq_tx_ready(dev)) {
+		struct uart_data *buf = k_fifo_get(sd->fifo, K_NO_WAIT);
+		u16_t written = 0;
+
+		/* Nothing in the FIFO, nothing to send */
+		if (!buf) {
+			uart_irq_tx_disable(dev);
+			return;
+		}
+
+		while (buf->len > written) {
+			written += uart_fifo_fill(dev,
+						  &buf->buffer[written],
+						  buf->len - written);
+		}
+
+		while (!uart_irq_tx_complete(dev)) {
```

Comment:
I'd be cautious about busy looping within IRQ handler, but if it's proven to work, I guess there's no harm in this particular case.

Reply:
Yup, unless flow control (which is disabled here) is blocking the transmission, there is basically nothing that can stop the UART hardware. Time spent in the loop is thus upward limited by the UART buffer length / baud rate.